src/object-store/src/layers/lru_cache/read_cache.rs

```diff
@@ -0,0 +1,267 @@
+                let size = {
+                    let stat = self.file_cache.stat(&read_key, OpStat::default()).await?;
+
+                    stat.into_metadata().content_length()
+                };
+
+                increment_gauge!(OBJECT_STORE_LRU_CACHE_ENTRIES, 1.0);
+                increment_gauge!(OBJECT_STORE_LRU_CACHE_BYTES, size as f64);
+                self.mem_cache
+                    .insert(read_key, ReadResult::Success(size as u32))
+                    .await;
+            }
+        }
+
+        Ok(self.stat().await)
+    }
+
+    /// Returns true when the read cache contains the specific file.
+    pub(crate) async fn contains_file(&self, path: &str) -> bool {
+        self.mem_cache.contains_key(path)
+            && self.file_cache.stat(path, OpStat::default()).await.is_ok()
+    }
+
+    /// Read from a specific path using the OpRead operation.
+    /// It will attempt to retrieve the data from the local cache.
+    /// If the data is not found in the local cache,
+    /// it will fallback to retrieving it from remote object storage
+    /// and cache the result locally.
+    pub(crate) async fn read<'life0, 'life1, 'async_trait, I, F>(
+        &'life0 self,
+        path: &'life1 str,
+        args: OpRead,
+        inner_read: F,
+    ) -> Result<(RpRead, Box<dyn Read>)>
+    where
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+        I: Accessor,
+        F: FnOnce(
+            &'life1 str,
+            OpRead,
+        ) -> Pin<
+            Box<dyn Future<Output = Result<(RpRead, I::Reader)>> + Send + 'async_trait>,
+        >,
+    {
+        let read_key = read_cache_key(path, &args);
+        let read_result = self.mem_cache.get(&read_key).await;
+
+        if let Some(read_result) = read_result {
+            let cache_result = match read_result {
+                ReadResult::Success(_) => {
+                    increment_counter!(OBJECT_STORE_LRU_CACHE_HIT, "result" => "success");
+
+                    // There is a concurrent issue here, the local cache may be purged
+                    // while reading, we have to fallback to remote read
+                    match self.file_cache.read(&read_key, OpRead::default()).await {
+                        Ok(ret) => Ok(to_output_reader(ret)),
+                        Err(_) => inner_read(path, args.clone()).await.map(to_output_reader),
+                    }
+                }
+                ReadResult::NotFound => {
+                    increment_counter!(OBJECT_STORE_LRU_CACHE_HIT, "result" => "not_found");
+
+                    Err(OpendalError::new(
+                        ErrorKind::NotFound,
+                        &format!("File not found: {path}"),
+                    ))
+                }
+            };
+            self.mem_cache.run_pending_tasks().await;
+
+            return cache_result;
+        }
+
+        increment_counter!(OBJECT_STORE_LRU_CACHE_MISS);
+
+        let inner_result = inner_read(path, args.clone()).await;
```

Comment:
issue: There might be some concurrent issues. e.g. One possible case
1. A key is evicted, we read the old file from the remote
2. A concurrent write updates the file in the remote and evicts the cache (the key doesn't exist so we have nothing to remove from the cache
3. We update the cache with old contents


I'm not sure whether `try_get_with` can solve this.
https://docs.rs/moka/latest/moka/future/struct.Cache.html#method.try_get_with

Reply:
Good catch!
The `try_get_with` can't solve this issue, because the write is in another thread exuection. If we want to solve this issue, we have to lock with some keys to ensure that read/write every cache item is in the same locking guard and can't run concurrently.

Reply:
Another solution: we can add a version to the cache item.  When invalidating the cache item, we increase the version field. 

When the read operation is issued to read remote file, it keeps the current version value and must ensure the version is not changed before updating the cache.

Reply:
This issue is the same as the cache consistency problem https://betterprogramming.pub/how-to-ensure-the-consistency-between-redis-and-database-62f09de0bdde

But we can't take these solutions here, we want strong consistency, not the final consistency.
So I will change back to the use of an allow list that determines if the file will be cached.